### PR TITLE
feat: support header parameters

### LINKED ISSUE #255: Headers not supported
https://github.com/oxidecomputer/progenitor/blob/4e2dcc5/progenitor-impl/src/method.rs#L330
e.g. I have a header which allows telling the backend the API version the client is expecting.
```yaml
  parameters:
    APIVersionParameterHeader:
      name: X-Foo-ApiVersion
      in: header
      required: false
      schema:
        type: string
```

### PR DESCRIPTION
**Pull request by DDtKey**
Closes #255
Base support for Header parameters.

And small additional change: use `Vec::with_capacity` instead of `Vec::new`

All current tests passes, but it would be nice to add some test with `header` params (I tested it on my own examples). 
I've not found `Header` in `dropshot` to write "specific-test". So we can choose any public openapi spec to add. (Upd: see https://github.com/oxidecomputer/progenitor/pull/210#issuecomment-1298794146)

### CODE REVIEW

**Comment by DDtKey** [comment]
@ahl could you take a look?  (sorry for bothering, i don't know who is maintainer)

**Comment by ahl** [comment]
Thanks for this. I'm going to take a look, but I wanted to set the expectation that it might be a week or two as I'm a little slammed at the moment.

**Comment by jayvdb** [comment]
IMO it would be helpful to add a header to the files in https://github.com/oxidecomputer/progenitor/tree/main/sample_openapi , esp `keeper.json`, and update the https://github.com/oxidecomputer/progenitor/tree/main/example-build if required to make this work.

**Comment by DDtKey** [comment]
@jayvdb I totally agree, but I didn't want to change existed APIs because I'm not sure if they really has any headers as params in their original specs. 

I've checked and found that `keeper` has required `Authorization` header ([link](https://github.com/jclulow/keeper/blob/b32a1f39ed1c54123662cf405e8e0ab3299fc8d2/server/src/main.rs#L69)), so I've added it to samples and updated tests.

**Comment by jayvdb** [comment]
fyi, for non-mandatory headers, I see

```rust
rror[E0277]: `?` couldn't convert the error to `progenitor_client::Error`
   --> clients/foo/src/lib.rs:501:78
    |
501 |             header_map.append("X-Foo-ApiVersion", HeaderValue::try_from(v)?);
    |                                                                              ^ the trait `std::convert::From<reqwest::header::InvalidHeaderValue>` is not implemented for `progenitor_client::Error`
    |
    = note: the question mark operation (`?`) implicitly performs a conversion on the error value using the `From` trait
    = help: the following other types implement trait `std::convert::From<T>`:
              <progenitor_client::Error<E> as std::convert::From<reqwest::Error>>
    = note: required for `std::result::Result<progenitor_client::ResponseValue<types::ApplicationRegistrationInfo>, progenitor_client::Error>` to implement `std::ops::FromResidual<std::result::Result<std::convert::Infallible, reqwest::header::InvalidHeaderValue>>`
```

**Review by jayvdb — APPROVED** [review]
WFM. Thanks.

**Comment by DDtKey** [comment]
Gentle ping @ahl

**Comment by DDtKey** [comment]
Gentle ping @ahl

**Review by ahl — COMMENTED** [review]
looks good; apologies for the long delay

File: progenitor-impl/src/method.rs
Comment (ahl) [inline progenitor-impl/src/method.rs:774]: I think this stood for "query name", but this is fine...

**Review by ahl — CHANGES_REQUESTED** [review]
can you please add a note to the CHANGELOG.md file?

**Comment by DDtKey** [comment]
> can you please add a note to the CHANGELOG.md file?

Well, I see only [this file](https://github.com/oxidecomputer/progenitor/blob/main/CHANGELOG.adoc) and it tells that it will be updated automatically on release. Isn't? 🤔

File: progenitor-impl/src/method.rs
Comment (DDtKey) [reply progenitor-impl/src/method.rs]: Oh, yeah. Changed to `hn` to keep the naming style

**Comment by ahl** [comment]
> > can you please add a note to the CHANGELOG.md file?
> 
> Well, I see only [this file](https://github.com/oxidecomputer/progenitor/blob/main/CHANGELOG.adoc) and it tells that it will be updated automatically on release. Isn't? 🤔

The sections are updated; the content is added by hand; check out the history for the file.

**Comment by DDtKey** [comment]
> > > can you please add a note to the CHANGELOG.md file?
> 
> > 
> 
> > Well, I see only [this file](https://github.com/oxidecomputer/progenitor/blob/main/CHANGELOG.adoc) and it tells that it will be updated automatically on release. Isn't? 🤔
> 
> 
> 
> The sections are updated; the content is added by hand; check out the history for the file.

Oh, got it. Check please.
A bit confusing to be honest, probably only to me.

**Review by ahl — APPROVED** [review]
nice

**Comment by ahl** [comment]
`EXPECTORATE=overwrite cargo test`

**Comment by DDtKey** [comment]
> `EXPECTORATE=overwrite cargo test`

Yeah, sorry for some mess here. I'm just not near to my main laptop. So I don't have all of tools. 

It should be fine now, just pls use squashing merge strategy if CI is green again 😅

**Comment by DDtKey** [comment]
@ahl updated, all checks are passed locally (was finally able to return home)

Sorry for the mess one more time

**Comment by DDtKey** [comment]
@ahl could you approve CI runs one more time? I believe we can merge this PR after last changes

**Comment by ahl** [comment]
rerunning a failed macos CI run that appeared to be a transient issue

**Comment by ahl** [comment]
woot! Thanks for taking the long slow walk to get this done @DDtKey!

**Pull request merged**
